cpp/harvest/core.cc

```diff
@@ -318,17 +318,46 @@ void Filter(int argc, char **argv) {
     // std::sort(filter_data_provider_ids.begin(), filter_data_provider_ids.end());
 
     for (auto work : works) {
-        if (not filter_data_provider_ids.empty()) {
-            auto sortedDataProvider = work.getDataProviderIds();
-            const bool is_in(MiscUtil::Intersect(sortedDataProvider, filter_data_provider_ids).size() > 0);
-            if (is_in == filter) {
-                work.setFilteredReason("Data Provider");
-                CORE::OutputFileAppend(output_file_skip, work, skipped == 0);
-                ++skipped;
-                ++skipped_data_provider_count;
-                continue;
+        if (argc == 7) {
+            if (!filter) {
+                // this means keep the record but clean the member of data provider
+                // clean the member of data provider means delete the member that its id is not in the list
+                // this is 'keep' option
+                auto data_provider_ids = work.getDataProviderIds();
+                const bool is_in_the_list(MiscUtil::Intersect(data_provider_ids, filter_data_provider_ids).size() > 0);
+                if (!data_provider_ids.empty() && is_in_the_list) {
+                    std::vector<nlohmann::json> new_data_provider;
+                    const auto data_providers = work.getDataProvider();
```

Comment:
This logic should be moved to CORE.h and CORE.cc => there should be a function called `removeDataProvider(unsigned long data_provider_id)` that should do this inside the library. We do not want to use nlohmann::json outside of CORE.h and CORE.cc.